test/soak/serve_hostnames/serve_hostnames.go

```diff
@@ -0,0 +1,202 @@
+		fmt.Printf("%d: %s\n", i, node.Name)
+	}
+
+	// Make a unique namespace for this test.
+	r := rand.New(rand.NewSource(time.Now().UnixNano()))
+	ns := "serve-hostnames-" + strconv.Itoa(r.Int()%10000)
+	fmt.Printf("Using namespace %s for this test.\n", ns)
+
+	// Put serve-hostname pods on each node.
+	podNames := []string{}
+	for i, node := range nodes.Items {
+		for j := 0; j < *perNode; j++ {
+			podName := fmt.Sprintf("serve-hostname-%d-%d", i, j)
+			podNames = append(podNames, podName)
+			fmt.Printf("Creating pod %s/%s on node %s ...", ns, podName, node.Name)
+			_, err := c.Pods(ns).Create(&api.Pod{
+				ObjectMeta: api.ObjectMeta{
+					Name: podName,
+					Labels: map[string]string{
+						"name": "serve-hostname",
+					},
+				},
+				Spec: api.PodSpec{
+					Containers: []api.Container{
+						{
+							Name:  "serve-hostname",
+							Image: "kubernetes/serve_hostname:1.1",
+							Ports: []api.ContainerPort{{ContainerPort: 9376}},
+						},
+					},
+					Host: node.Name,
+				},
+			})
+			if err != nil {
+				log.Fatalf("Failed to create pod server-hostname-%d", i)
+			}
+			fmt.Print(" [done]\n")
+		}
+	}
+	// Clean up the pods
+	defer func() {
+		for _, podName := range podNames {
+			if err = c.Pods(ns).Delete(podName); err != nil {
+				log.Fatalf("Failed to delete pod %s: %v", podName, err)
+			}
+		}
+	}()
+
+	// Create a service for these pods.
+	fmt.Print("Creating service serve-hostnames ...")
+	svc, err := c.Services(ns).Create(&api.Service{
+		ObjectMeta: api.ObjectMeta{
+			Name: "serve-hostnames",
+			Labels: map[string]string{
+				"name": "serve-hostname",
+			},
+		},
+		Spec: api.ServiceSpec{
+			Port:       9376,
+			TargetPort: util.NewIntOrStringFromInt(9376),
+			Selector: map[string]string{
+				"name": "serve-hostname",
+			},
+		},
+	})
+	if err != nil {
+		log.Fatalf("unable to create service %s/%s: %v", ns, svc.Name, err)
+	}
+	fmt.Print(" [done]\n")
+	// Clean up service
+	defer func() {
+		if err = c.Services(ns).Delete(svc.Name); err != nil {
+			log.Fatalf("unable to delete service %s/%s: %v", ns, svc.Name, err)
+		}
+	}()
+
+	fmt.Print("Wait for the serve-hostname pods to be ready\n")
+	for _, podName := range podNames {
+		var pod *api.Pod
+		for start := time.Now(); time.Since(start) < podStartTimeout; time.Sleep(5 * time.Second) {
+			pod, err = c.Pods(ns).Get(podName)
+			if err != nil {
+				fmt.Printf("Get pod %s/%s failed, ignoring for %v: %v\n", ns, podName, err, podStartTimeout)
+				continue
+			}
+			if pod.Status.Phase == api.PodRunning {
+				break
+			}
+		}
+		if pod.Status.Phase != api.PodRunning {
+			log.Fatalf("Gave up waiting on pod %s/%s to be running (saw %v)", ns, podName, pod.Status.Phase)
+		}
+		fmt.Printf("%s/%s is running\n", ns, podName)
+	}
+
+	// Wait a bit for the service to notice.
+	fmt.Print("Waiting for the endpoints to propagate... ")
+	time.Sleep(30 * time.Second)
+	fmt.Printf("finished waiting\n")
+
+	// Repeatedly make requests.
+	for {
```

Comment:
Added a flag to specify the number of iterations. A (default) negative value means no iterations.